src/caustics/utils.py

```diff
@@ -115,6 +116,124 @@ def get_meshgrid(
     )
     return torch.meshgrid([xs, ys], indexing="xy")
 
+@lru_cache(maxsize=32)
+def _quad_table(n, p, dtype, device):
+    """
+    Generate a meshgrid for quadrature points using Legendre-Gauss quadrature.
+
+    Parameters
+    ----------
+    n : int
+        The number of quadrature points in each dimension.
+    p : torch.Tensor
+        The pixelscale.
+    dtype : torch.dtype
+        The desired data type of the tensor.
+    device : torch.device
+        The device on which to create the tensor.
+
+    Returns
+    -------
+    Tuple[torch.Tensor, torch.Tensor, torch.Tensor]
+        The generated meshgrid as a tuple of Tensors.
+    """
+    abscissa, weights = roots_legendre(n)
+
+    w = torch.tensor(weights, dtype=dtype, device=device)
+    a = p * torch.tensor(abscissa, dtype=dtype, device=device) / 2.0
+    X, Y = torch.meshgrid(a, a, indexing="xy")
+
+    W = torch.outer(w, w) / 4.0
+
+    X, Y = X.reshape(-1), Y.reshape(-1) # flatten
+    return X, Y, W.reshape(-1)
+
+def get_pixel_quad_integrator_grid(
+    pixelscale, 
+    X, 
+    Y, 
+    quad_level = 3, 
+    device=None, 
+    dtype=torch.float32,
+):
+    """
+    Generates a 2D meshgrid for Gaussian quadrature based on the provided pixelscale and dimensions.
+
+    Parameters
+    ----------
+    pixelscale : float
+        The scale of the meshgrid in each dimension.
+    X : Tensor
+        The x-coordinates of the pixel centers.
+    Y : Tensor
+        The y-coordinates of the pixel centers.
+    quad_level : int, optional
+        The number of quadrature points in each dimension. Default is 3.
+    device : torch.device, optional
+        The device on which to create the tensor. Default is None.
+    dtype : torch.dtype, optional
+        The desired data type of the tensor. Default is torch.float32.
+
+    Returns
+    -------
+    Tuple[Tensor, Tensor]
+        The generated meshgrid as a tuple of Tensors.
+
+    Example
+    -------
+    Usage would look something like:: python
+
+        X, Y = get_meshgrid(pixelscale, nx, ny)
+        Xs, Ys, weight = get_pixel_quad_integrator_grid(pixelscale, X, Y, quad_level)
+        F = your_brightness_function(Xs, Ys, other, parameters)
+        res = pixel_quad_integrator(F, weight)
+    """
+
+    # collect gaussian quadrature weights
+    abscissaX, abscissaY, weight = _quad_table(
+        quad_level, pixelscale, dtype, device
+    )
+
+    # Gaussian quadrature evaluation points
+    Xs = torch.repeat_interleave(X[..., None], quad_level ** 2, -1) + abscissaX
+    Ys = torch.repeat_interleave(Y[..., None], quad_level ** 2, -1) + abscissaY
+    return Xs, Ys, weight
+
+def pixel_quad_integrator(
```

Comment:
Oh I see, so you see the Gaussian quadrature integrator as a replacement of something like average pooling? Currently we do not have any model that work on irregular pixel grids, and so I am wondering what this brings. Is it a better integrator for sharp sources like the Sersic model?

In that case my second comment is wrong as you pointed out, this definitely belongs in the simulator at the same level as average pooling.

Reply:
Yes, it is basically average pooling that converges faster. Average pooling is basically integrating with the midpoint rule. One step better is simpsons rule, and two steps better is gaussian quadrature :)

Reply:
Wait I missed something, we have an upsampling layer in there? Ideally, we would be evaluating the source model at the same resolution as the PSF so that only downsampling is performed in the forward model. Unless the user specifies a specific pixel size for a pixelated source model, but then the situation is a bit more complicated. 

And I am all for the gaussian_quadrature setup, as long as the default is average_pooling. I care about this in a somewhat pedantic fashion because of the linearity of the simulator with respect to the source parameters (for pixelated models).

Reply:
Ok let's separate two things. For a pixelated model the situation is a bit complicated. But for Sersic, you basically use the Gaussian quadrature to pixelate the model at the PSF resolution is that correct? But then I feel like we should have average pooling to go back to the observation resolution

Reply:
Ok I went back to read the full code and this is what is going on: pixelate at PSF resolution and then convolve and then pixelate at observation resolution. Sorry my bad I was a bit slow here. 

But then my question about the pixelated source remains. How do you handle this case?

Reply:
Right, the upsampling should only be used if the PSF is upsampled, otherwise it is better to use quadrature for the accuracy/evaluation ratio. The way the code stands right now, the user can choose very freely between the two. Here are the valid configurations:

- no PSF convolution, user can freely choose upsample factor, or quadrature factor, or both at will 
- with PSF convolution, user should set upsample factor to the upsampling of the PSF, quadrature can be set to any level

So the quadrature can always be set to anything and it is valid. The upsampling/avgpooling combo has to match the PSF if there is a PSF. 

As for linearity, the gaussian quadrature is also linear. It is actually equivalent to a convolution with stride equal to the kernel size, which is the same as kernelized block pooling which is what I do.

Reply:
A pixelated source is treated like any other source. Since a pixelated source is defined by bilinear interpolation, it can also be evaluated an integrated at any set of coordinates. That said, because a pixelated source is linear at the pixel scale, there is no reason to do sub-pixel integration as the midpoint rule is exact for linear integrals.

Reply:
Ok make sense. Thanks for the explanation!

Reply:
This is some good discussion. I'll add a rewritten version to the documentation :)